### ISSUE TITLE
Confusing error message for "Can't solve constraint"

### ISSUE THREAD

**Issue by farnoy** [idris-lang/Idris2#945]
Hi, I have two problems with the code posted below. The first one is that I'm not sure where the constraint is coming from, the second one with the compiler not showing me how it arrived at this constraint.

The problem I'm solving is to create a data type that associates natural numbers with the power of two needed to represent them. You can think of this as "how many bits to store n".

# Steps to Reproduce

```idris
module Repro

import Data.Nat

%default total

pow2 : Nat -> Nat
pow2 n = power 2 n

plusOneLTEPlusSelf : (a : Nat) -> GT a 0 -> LTE (S a) (a + a)
plusOneLTEPlusSelf 0 _ impossible
plusOneLTEPlusSelf (S n) _ =
  let y = lteAddRight (S n) {m=n} in
    rewrite sym (plusSuccRightSucc n n) in
      LTESucc y

greaterPowerStillGreater : (a, b : Nat) -> GT a 0 -> a = pow2 b -> LTE (S a) (pow2 (S b))
greaterPowerStillGreater a b gt0 prf =
  rewrite sym prf in
    rewrite plusZeroRightNeutral a in
      plusOneLTEPlusSelf a gt0

data NextPowerOfTwo : (n, p : Nat) -> { auto proof : LTE n (pow2 p) } -> Type where
  NextPowZero : NextPowerOfTwo 0 0
  NextPowSucc : NextPowerOfTwo a b {proof=renamed}
             -> { auto p : LTE (S a) (pow2 b) }
             -> NextPowerOfTwo (S a) b
  NextPowSuccPow : NextPowerOfTwo a b {proof=renamed}
                -> { auto p : a = pow2 b }
                -> { auto gt0 : GT a 0 }
                -> NextPowerOfTwo (S a) (S b) {proof=greaterPowerStillGreater a b gt0 p}

test0 : NextPowerOfTwo 0 0
test0 = NextPowZero

test1 : NextPowerOfTwo 1 0
test1 = NextPowSucc test0

test2 : NextPowerOfTwo 2 1
test2 = NextPowSuccPow test1

test3 : { auto k : NextPowerOfTwo 2 1 } -> NextPowerOfTwo 3 2
test3 {k} = NextPowSuccPow k
```

# Expected Behavior

I expected this to compile & work and for the compiler to be able to proof search my data structure to provide the `auto` implicits.

# Observed Behavior

Errors whenever I try to use `NextPowSuccPow`:
```
Error: While processing right hand side of test2. Can't solve constraint
between: lteAddRight 1 and LTESucc LTEZero.

Repro.idr:40:9--40:29
    |
 40 | test2 = NextPowSuccPow test1
    |         ^^^^^^^^^^^^^^^^^^^^

Error: While processing right hand side of test3. Can't solve constraint
between: lteAddRight 2 and LTESucc (LTESucc LTEZero).

Repro.idr:43:13--43:29
    |
 43 | test3 {k} = NextPowSuccPow k
    |             ^^^^^^^^^^^^^^^^
```

-----

Through trial and error, I figured out that dropping the implicit `proof` in the data declaration resolved this problem with no downsides that I can observe. I still don't understand why it caused problems and if my representation can be sound without it.

<details>
<summary>The fixed version with `proof` removed:</summary>

```idris
module Repro2

import Data.Nat

%default total

pow2 : Nat -> Nat
pow2 n = power 2 n

data NextPowerOfTwo : (n, p : Nat) -> Type where
  NextPowZero : NextPowerOfTwo 0 0
  NextPowSucc : NextPowerOfTwo a b -> { auto p : LTE (S a) (pow2 b) } -> NextPowerOfTwo (S a) b
  NextPowSuccPow : NextPowerOfTwo a b
                -> { auto p : a = pow2 b }
                -> { auto gt0 : GT a 0 }
                -> NextPowerOfTwo (S a) (S b)

test0 : NextPowerOfTwo 0 0
test0 = NextPowZero

test1 : NextPowerOfTwo 1 0
test1 = NextPowSucc test0

test2 : NextPowerOfTwo 2 1
test2 = NextPowSuccPow test1

test3 : { auto k : NextPowerOfTwo 2 1 } -> NextPowerOfTwo 3 2
test3 {k} = NextPowSuccPow k
```
</details>

**Comment by Russoul** [comment]
Adding this before `pow2` fixes your problem:
```
%hide Data.Nat.lteAddRight

lteAddRight : (n : Nat) -> LTE n (n + m)
lteAddRight Z = LTEZero
lteAddRight (S k) {m} = LTESucc (lteAddRight {m} k)

```

The error you are experiencing is due to the access modifier of `lteAddRight` declared in `Data.Nat` being too restrictive.
It is declared as `export` meaning that its definition (pattern matching clauses) can't be discovered and used in any file other than the one it is declared in.

That's why you get `Can't solve constraint
between: lteAddRight 1 and LTESucc LTEZero`.
Idris can't reduce `lteAddRight 1` to `LTESucc LTEZero`. We can get around the problem by duplicating the definition.
Which is not ideal... I kind of still don't understand why we have such behaviour. You can make a PR to change the `export` flag to `public export` to make it work as expected and for others to enjoy compile time reduction :).

**Comment by farnoy** [comment]
Thanks for the help, I'm relieved that it was something simple! I'll prepare a PR because it seems useful. I think you meant compile _speed_ reduction, right? Changing from `export` to `public export` probably adds work for the compiler.

By the way, is it possible to make this function work?

```idris
howManyBitsFor : (a : Nat) -> { auto bits : Nat } -> { auto np : NextPowerOfTwo a bits } -> Nat
howManyBitsFor _ {bits} = bits
```

It typechecks but can't find the implicits on its own. Proof search only seems to work when the parameters are known in advance, even though I was careful to design a data structure having at most one representation for each pair of `(a, b : Nat)`

**Comment by gallais** [comment]
> Proof search only seems to work when the parameters are known in advance, even though I was careful to design a data structure having at most one representation for each pair of `(a, b : Nat)`

It's often a bad idea to rely on `auto` for anything that is non-trivial
(and things can be unique but still non-trivial).

If you know how to compute a thing then it's better to write that function
yourself rather than hoping that `auto` will figure out the right output
each time it is fed an input. It's a lot easier for Idris to run a computation
than to explore an arbitrary search space.

**Comment by gallais** [comment]
Fixed in #948
